### ISSUE TITLE
Operating System install not recognised for Wkhtmltopdf option in 1.9.0

### ISSUE THREAD

**Issue by hughbris** [iusvar/grav-plugin-snappygrav#44]
This has only happened since I upgraded to v1.9.0.

I had previously installed wkhtmltopdf on the O/S on my server (`/usr/bin`) because of problems getting a composer-based install to work. I configured 'position' to O/S and the path was already set correctly. When generating a PDF, the plugin told me I need to install wkhtmltopdf and provided the composer commands. Scanning through the PHP class code, it looks like the 'position' field is not checked before showing this dialogue.

_(However I was happy to see that the other two PDF library options installed correctly and generated PDFs in on my server in v1.9.0. I just decided to see if wkhtmltopdf worked in the new version, but it hopefully doesn't matter :) )_

**Comment by iusvar** [comment]
Actually I check the existence of the libraries by checking in the composer.json file.
And this is not enough when wkhtmltopdf is already installed in the operating system, as I had foreseen.
So I anticipated a further check when the selected library is wkhtmltopdf and its position is the operating system.

(_You're right, the presence of more libraries is relaxing, although they obviously offer different results in terms of speed of execution, quality in the transformation from html to pdf, ..._)

**Comment by hughbris** [comment]
Out of interest, which of the new libraries do you recommend? I am planning a complex layout that emulates a populated existing paper form (within Data Manager), with:

* columns
* boxes for sections
* checkboxes, some with checkmarks
* lines
* some colour
* a logo
* possibly a simple generated vector graphic
* maybe embedded images (but it might be better to serve them as separate hi-res attachments).

Features are probably more important than speed. I guess I should start by trying to reproduce the populated form in a Twig template.

**Comment by iusvar** [comment]
The answer is not simple because all three libraries allow you to manage the html elements / objects you list.

The difference is in the way they are managed them: for some it is enough that they are present in the html string to be treated (hence the output twig), for others it is necessary to manage them directly in php.

In general there are no big differences though, lately I'm preferring TCPDF. In the future, as I add new options, I will evaluate them all fairly to realize the differences.

From what you write, however, I understand that you want to fill in the pdf forms (probably already existing?) With data from Data Manager, am I wrong?

If I'm right I tell you that maybe it would be better to think of something else, for example at **pdftk**. I myself am doing some tests after reading this  [dialogue](https://discordapp.com/channels/501836936584101899/506916956637495306/532912977645404161), on January 10, 2019, in the chat between **@Gravabal** and **@Ricardo** (but the material still unformed is not yet available on github)

**Comment by hughbris** [comment]
Yes, the PDF "forms" (they only have to look like forms) are populated from data sourced from existing public HTML forms ([[1]](https://qe2dental.co.nz/forms/medical-history), [[2]](https://qe2dental.co.nz/forms/practitioner-referral)) and Data Manager is the logical secure place to access them. The first one needs to be printed and signed on premises, then stored as an official record (scanned or physical, up to the client). Both are legal documents.

Thanks to your `servePDF` method (although its name is not quite accurate!), it's very simple to generate PDFs from outside the plugin now and experiment with the different engines :)

So hopefully I'll give you some feedback on what works soon.
